### PR TITLE
Create Lazy-Loaded Twig Extension

### PR DESCRIPTION
**Pull request by emmanuelballery**
| Q | A
| --- | ---
| Branch? | 2.x
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Tests pass? | yes
| Fixed tickets | #1261
| License | MIT
| Doc PR | 

Hi!

This is a new version for #1265 ;

Twig supports [Lazy-Loaded Extensions](https://symfony.com/doc/current/templating/twig_extension.html#creating-lazy-loaded-twig-extensions) since `1.35.0` and `2.4.4`.

Tests have been updated too.

(Also, @mbabker provided details and Blackfire comparison in #1261)

### CODE REVIEW

File: Templating/FilterExtension.php
Comment (emmanuelballery) [inline Templating/FilterExtension.php:19]: This may be a BC since `FilterExtension` is not `@internal` and people may have extended it - a strange thing to do. But it's also the only way to prevent Symfony container to load the `CacheManager` when it's not necessary.

File: Tests/Templating/FilterRuntimeTest.php
Comment (dmaicher) [inline Tests/Templating/FilterRuntimeTest.php]: This cannot happen I think? twig is a hard dependency and the supported versions should have this interface

**Comment by emmanuelballery** [comment]
Sorry @dmaicher just checked again and this question was rulled out here https://github.com/sulu/sulu/issues/5556#issuecomment-832932399_ in someone else attempt to stop loading this extension.

Feel free to close then, I'll disable this extension per project I guess.

**Comment by dmaicher** [comment]
@emmanuelballery I'm not a member on this repo :wink: I just stumbled across your PR as I also noticed the non-lazy twig extension and all the services that are instantiated although mostly not used in my case. So I would like to see this implemented :+1: 

I believe this would need to go to  `3.x` as its for sure a BC break.

**Comment by lsmith77** [comment]
Maybe we can figure out a non BC breaking approach and then I would be fine to have something like this in 2.x

/cc @fbourigault

**Comment by dmaicher** [comment]
> Maybe we can figure out a non BC breaking approach and then I would be fine to have something like this in 2.x
> 
> /cc @fbourigault

One way of achieving this with BC would be to
- create a new Extension class like `FilterRuntimeExtension` or `LazyFilterExtension` 
- deprecate the current `FilterExtension` and the `FilterTrait` 
- add a config parameter `liip_imagine.twig_runtime_extension: false` that is false by default and deprecate not setting it to true (+ we could set it to true on the [recipe](https://github.com/symfony/recipes-contrib/blob/master/liip/imagine-bundle/1.8/config/packages/liip_imagine.yaml#L2))
- register the old extension if its false and the new one if its true

Then on 3.0 we could remove the old one and always use the lazy one with the runtime.

Or does someone have a better idea?

**Comment by emmanuelballery** [comment]
I'd:
- provide the new `LazyFilterExtension`
- deprecate the rest
- give people the choice whether they want to move on by registering the new service themselves (overriding the existing one - this way we don't add extra options to remove for 3.x)

```yaml
liip_imagine.templating.filter_extension:
    class: Liip\ImagineBundle\Templating\LazyFilterExtension
    tags: [ 'twig.extension' ]
```

Then in 3.x register `LazyFilterExtension` ourselves ;

**Comment by lsmith77** [comment]
both approaches sound fine for me.

FYI if you rebase, PHPCS fixer should work again.

File: Resources/config/imagine.xml
Comment (emmanuelballery) [inline Resources/config/imagine.xml]: Not sure if I had to do this.

In 2.x people can override this themselves:

```yaml
    liip_imagine.templating.filter_extension:
        class: Liip\ImagineBundle\Templating\LazyFilterExtension
        tags: [ 'twig.extension' ]
    liip_imagine.templating.filter_runtime:
        class: Liip\ImagineBundle\Templating\FilterRuntime
        arguments: [ '@liip_imagine.cache.manager' ]
        tags: [ 'twig.runtime' ]
```

File: Templating/LazyFilterExtension.php
Comment (dmaicher) [inline Templating/LazyFilterExtension.php:36]: I think we are missing

```php
    /**
     * {@inheritdoc}
     */
    public function getName()
    {
        return 'liip_imagine';
    }
```

or?

File: Templating/FilterExtension.php
Comment (dmaicher) [inline Templating/FilterExtension.php]:
```suggestion
@trigger_error('The ' . FilterExtension::class . ' class is deprecated since version 2.7 and will be removed in 3.0; use ' . LazyFilterExtension::class . ' instead.', E_USER_DEPRECATED);
```

?

File: Templating/FilterTrait.php
Comment (dmaicher) [inline Templating/FilterTrait.php]: we should also add `@deprecated` class annotations for both I think

**Review by dmaicher — COMMENTED** [review]
I would have preferred the other approach with a simpler upgrade path for the users but this is also fine for me :+1:

File: Templating/LazyFilterExtension.php
Comment (dmaicher) [inline Templating/LazyFilterExtension.php]: I think we should make this class final. There is no point in extending it

File: Templating/LazyFilterExtension.php
Comment (emmanuelballery) [reply Templating/LazyFilterExtension.php:36]: Missed the fact that Twig 1 was still supported :/

**Review by dmaicher — APPROVED** [review]

File: Templating/FilterTrait.php
Comment (emmanuelballery) [reply Templating/FilterTrait.php]: I followed what was done in the other files, but I agree with you.

**Comment by emmanuelballery** [comment]
@dmaicher Maybe I did not completely understand your first suggestion. I was under the impression you were asking to add a configuration and that this configuration would then be obsolete in `3.x` (more work for this bundle maintainers). I chose this approach because IMO only advanced users are interested in preventing Twig from loading more container's stuff than necessary.

- The upgrade path for this bundle is only to delete what's deprecated in `3.x` and uncoment the new config ;
- The upgrade path for users is to override the config now and remove (if they want) their configuration after :

```yaml
    liip_imagine.templating.filter_extension:
        class: Liip\ImagineBundle\Templating\LazyFilterExtension
        tags: [ 'twig.extension' ]
    liip_imagine.templating.filter_runtime:
        class: Liip\ImagineBundle\Templating\FilterRuntime
        arguments: [ '@liip_imagine.cache.manager' ]
        tags: [ 'twig.runtime' ]
```

I'm completly ok to work on your approach if maintainers wants me to 👍🏻

**Comment by dmaicher** [comment]
> @dmaicher Maybe I did not completely understand your first suggestion. I was under the impression you were asking to add a configuration and that this configuration would then be obsolete in `3.x` (more work for this bundle maintainers). I chose this approach because IMO only advanced users are interested in preventing Twig from loading more container's stuff than necessary.
> 
>     * The upgrade path for this bundle is only to delete what's deprecated in `3.x` and uncoment the new config ;
> 
>     * The upgrade path for users is to override the config now and remove (if they want) their configuration after :
> 
> 
> ```yaml
>     liip_imagine.templating.filter_extension:
>         class: Liip\ImagineBundle\Templating\LazyFilterExtension
>         tags: [ 'twig.extension' ]
>     liip_imagine.templating.filter_runtime:
>         class: Liip\ImagineBundle\Templating\FilterRuntime
>         arguments: [ '@liip_imagine.cache.manager' ]
>         tags: [ 'twig.runtime' ]
> ```
> 
> I'm completly ok to work on your approach if maintainers wants me to 👍🏻

The downside of your approach is that users might see a deprecation notice since they are using the old filter. Then what do they do? They would need to find this PR (or check the CHANGELOG or so) to find out that they need to register a service themselves to fix it. Then once 3.0 is released they need to remember to remove that custom service definition again as it works without then.

So my idea was to make this a bit simpler. They get a deprecation because of the added config option. They set it to true and for 99.9% of all users this works fine. Only if someone extended the filter or so they need to rework it.
Then once 3.0 is released they get an error because of the removed config option. They remove the option and all works.
EDIT: also for new installations using the recipe we could immediately use the new filter runtime for those users (config option true)

But in the end its not for me to decide :blush: I'm fine with both options

**Comment by dmaicher** [comment]
@lsmith77 how can we move this forward? :blush: Can I help somehow?

**Comment by lsmith77** [comment]
@dbu / @Tobion

**Comment by dmaicher** [comment]
Friendly ping :blush: Can a maintainer review this?

File: Templating/FilterExtension.php
Comment (dbu) [reply Templating/FilterExtension.php:19]: with the current version of the PR, this is no BC break anymore, right?

File: Resources/config/imagine.xml
Comment (dbu) [reply Resources/config/imagine.xml]: i think it would be elegant to move this definition into its own file, and the above one for the old extension into a separate file too.

and add a configuration option to select which twig mode to use. (none/lazy/legacy, defaulting to legacy)

the deprecation note can then also tell the configuration option to switch to the lazy extension.

**Comment by dbu** [comment]
sorry for the long silence. i have no some time reserved each week to keep LiipImagineBundle alive.

i added some minor comments, but like where this is going. can you please have a look at them?

and can you please rebase the branch on 2.x to get the latest changes and ci-adjustments in?

File: Resources/config/imagine.xml
Comment (dmaicher) [reply Resources/config/imagine.xml]: yeah choosing the runtime extension via a config setting (+ trigger deprecation for legacy non-lazy way) would also be my preferred way

**Comment by emmanuelballery** [comment]
God ... I messed up 😞

**Comment by emmanuelballery** [comment]
Sorry for the noise ... @dmaicher is it what you were thinking of?

630016fb7a65ad937d5e51c25d5c6d8e77b61604

I am really rusty regarding bundle configuration...

File: DependencyInjection/Configuration.php
Comment (dmaicher) [inline DependencyInjection/Configuration.php:200]: What is the difference between `none` and  `legacy`? Wondering if we need a `none` mode?

Or should this be used if someone does not need the twig extension/runtime at all and we should not register any services then?

Comment (emmanuelballery) [reply DependencyInjection/Configuration.php:200]: I took this suggestion from @dbu's comment.
- `none` : no runtime, dependencies will be loaded with the container ;
- `lazy`: with runtime, dependencies will be loaded only when needed ;
- `legacy` is only here temporarily to inform users they now have the choice (triggering a deprecation) and will be removed in `3.0` - same as `none` regarding dependencies loading ;

Comment (dbu) [reply DependencyInjection/Configuration.php:200]: my idea of `none` was to not load the twig extension at all ;-)

is there a use case for loading the twig extension in legacy mode, that is without the lazy loading? (apart from having it until version 3 for BC, of course). if there is a use case, i'd call the modes none/lazy/eager and default to eager. 

but with the code deprecations, my understanding is that eager loading will be removed? then imho it would make sense to have none => do not load twig at all , lazy => what you did, legacy => the current legacy

**Comment by dbu** [comment]
can you please rebase on 2.x? that should improve the build situation...

i think the 8.0 test failure is because of our deprecation warnings. afaik we need to mark the tests for deprecated things with `@group legacy` to have phpunit not complain: https://symfony.com/doc/current/components/phpunit_bridge.html#mark-tests-as-legacy

File: DependencyInjection/Configuration.php
Comment (dmaicher) [reply DependencyInjection/Configuration.php:200]: Ok makes sense to me :blush: Let's keep `none/lazy/legacy` then and do not load any twig services with `none`. For the next major we can remove `legacy` then.

File: DependencyInjection/Configuration.php
Comment (franmomu) [inline DependencyInjection/Configuration.php]:
```suggestion
            ->enumNode('twig_mode')
                ->values('none', 'lazy', 'legacy')
```

We could use `enumNode` and there is no need of the first `validate` call.

**Review by dmaicher — APPROVED** [review]

**Review by dbu — APPROVED** [review]
great, i am happy with this now!

one last thing: can you please document the new twig_mode option in https://github.com/liip/LiipImagineBundle/blob/2.x/Resources/doc/configuration.rst?

**Comment by emmanuelballery** [comment]
I used this code in my projects and found some stupid mistakes ... sorry.

- `enumNode` in `Configuration` needed an `array` as parameters ;
- my `imagine_twig_mode_lazy.xml` was not referencing the new `Runtime` class ;
- I added documentation ;

**Review by dbu — APPROVED** [review]
awesome, thanks a lot!

**Comment by dbu** [comment]
thanks for having tested it. the mistake with the configuration would hopefully also have shown when github runs the tests. (it is always waiting for my confirmation in PRs from non repository members, afaik because some assholes started mining bitcoin on ci runners...)

can you please look at the cs fixer?

and try the `@group legacy` annotations?

> i think the test failure is because of our deprecation warnings. afaik we need to mark the tests for deprecated things with @group legacy to have phpunit not complain: https://symfony.com/doc/current/components/phpunit_bridge.html#mark-tests-as-legacy

**Comment by dbu** [comment]
cheers! i will look into releasing a new minor version hopefully this wednesday.

**Comment by emmanuelballery** [comment]
@dbu I may have broken the configuration doc ... 😞 I guess I had to add 2 spaces before each new line for it to work.

https://github.com/liip/LiipImagineBundle/blob/2.x/Resources/doc/configuration.rst

image

**Comment by dbu** [comment]
thanks for the heads up. fixing this in #1396

**Pull request merged**
